math/mathcore/inc/Fit/EvaluateLogL.hxx

```diff
@@ -0,0 +1,569 @@
+         std::cout << "\n===> Evaluate Gradient for parameters ";
+         for (unsigned int ip = 0; ip < npar; ++ip)
+            std::cout << "  " << p[ip];
+         std::cout << "\n";
+#endif 
+
+         (const_cast<IGradModelFunctionTempl<T> &>(func)).SetParameters(p);
+
+         const T kdmax1 = vecCore::math::Sqrt(vecCore::NumericLimits<T>::Max());
+         const T kdmax2 = vecCore::NumericLimits<T>::Max() / (4 * initialNPoints);
+
+         auto mapFunction = [&](const unsigned int i) {
+            std::vector<T> gradFunc(npar);
+            std::vector<T> pointContributionVec(npar);
+
+            T x1;
+            vecCore::Load<T>(x1, data.GetCoordComponent(i * vecSize, 0));
+
+            const T *x = nullptr;
+
+            unsigned int ndim = data.NDim();
+            std::vector<T> xc(ndim);
+            if (ndim > 1) {
+               xc.resize(ndim);
+               xc[0] = x1;
+               for (unsigned int j = 1; j < ndim; ++j)
+                  vecCore::Load<T>(xc[j], data.GetCoordComponent(i * vecSize, j));
+               x = xc.data();
+            } else {
+               x = &x1;
+            }
+
+
+            T fval = func(x, p);
+            func.ParameterGradient(x, p, &gradFunc[0]);
+
+#ifdef DEBUG_FITUTIL            
+            if (i < 5 || (i > numVectors-5) ) {
+               if (ndim > 1) std::cout << i << "  x " << x[0] << " y " << x[1] << " gradient " << gradFunc[0] << "  " << gradFunc[1] << "  " << gradFunc[3] << std::endl;
+               else std::cout << i << "  x " << x[0] << " gradient " << gradFunc[0] << "  " << gradFunc[1] << "  " << gradFunc[3] << std::endl;
+            }
+#endif            
+
+            vecCore::Mask<T> positiveValues = fval > 0;
+
+            for (unsigned int kpar = 0; kpar < npar; ++kpar) {
+               if (!vecCore::MaskEmpty(positiveValues))
+                  vecCore::MaskedAssign<T>(pointContributionVec[kpar], positiveValues, -1. / fval * gradFunc[kpar]);
+
+               vecCore::Mask<T> nonZeroGradientValues = !positiveValues && gradFunc[kpar] != 0;
+               if (!vecCore::MaskEmpty(nonZeroGradientValues)) {
+                  T gg = kdmax1 * gradFunc[kpar];
+                  pointContributionVec[kpar] =
+                     vecCore::Blend(nonZeroGradientValues && gg > 0, -vecCore::math::Min(gg, kdmax2),
+                                    -vecCore::math::Max(gg, -kdmax2));
+               }
+               // if func derivative is zero term is also zero so do not add in g[kpar]
+            }
+
+            return pointContributionVec;
+         };
+
+         // Vertically reduce the set of vectors by summing its equally-indexed components
+         auto redFunction = [&](const std::vector<std::vector<T>> &pointContributions) {
+            std::vector<T> result(npar);
+
+            for (auto const &pointContributionVec : pointContributions) {
+               for (unsigned int parameterIndex = 0; parameterIndex < npar; parameterIndex++)
```

Comment:
`s/parameterIndex/i/`? Why such long names for loop variables? It's strange to have this, and then `npar` is a short name...

Reply:
For clarity?